### ISSUE TITLE
Installation tokens are hard to re-use

### ISSUE THREAD

**Issue by marcusirgens** [XAMPPRocky/octocrab#435]
It is hard to re-use installation tokens, as it is not possible to mutate the auth state of the client, nor possible to extract the active installation token. Additionally, the installation auth state does not cache expiration times, installation ID or other data required to renew itself, so it times out after a while.

Would you be interested in a PR that changes any of these issues? My suggestions would be:
- Make it possible to retrieve the `InstallationToken` from the client, if present.
- Make it possible to set an `InstallationToken` on the client if the auth state is App
- Change the data structure for AuthState's installation variant to hold the expiration time and installation ID, and re-fetch an installation token if it is about to expire.

**Comment by XAMPPRocky** [comment]
> Change the data structure for AuthState's installation variant to hold the expiration time and installation ID, and re-fetch an installation token if it is about to expire (for octocrab-managed caching).

I think I would lean towards this, because ideally with a high-level client, this would be an implementation detail you shouldn't need to worry about.

**Comment by marcusirgens** [comment]
> I think I would lean towards this, because ideally with a high-level client, this would be an implementation detail you shouldn't need to worry about.

I agree, it would essentially behave like Go's OAuth2 client middleware in that case. It should be orthagonal with the two other points, ie., all three should be possible to do at the same time.

**Comment by flying-sheep** [comment]
Hi, I’m developing an app using installation tokens. It worked fine in the beginning, now trying to run the following

```rust
let base = Octocrab::builder().app(APP_ID, key).build()?;
let inst = base.apps().get_org_installation(ORG).await?;
```

errors with a `GitHubError`:

> 'Expiration time' claim ('exp') is too far in the future

I’m not sure if this PR would even fix that, since the error could have to do with clocks being desynchronized or so, but it might be related.
